### DISCUSSION TITLE
yolov5/tutorials/tips_for_best_training_results/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8777] [Docs]
# yolov5/tutorials/tips_for_best_training_results/

Our comprehensive guide provides insights on how to train your YOLOv5 system to get the best mAP. Master dataset preparation, model selection, training settings, and more.

https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/

**Comment by Halafz7** [comment]
It would be nice if you add tutorial how to train custom model (such as YOLOv3-quantized or something)

**Reply by glenn-jocher** [reply]
@Halafz7 I would recommend a visit to https://docs.ultralytics.com/modes/train for training code examples with YOLOv8 :)

**Comment by thomasjubin** [comment]
hello team,
 I am doing a project object dictection on road damages as part of my thesis and pretty new to this field, I have 4 classes and total of close to 6000 images including augmented images,  these are the results obtained after training using yoloV5s  for 64 epochs using google colab, 
YOLOv5s summary: 157 layers, 7020913 parameters, 0 gradients, 15.8 GFLOPs
                 Class     Images  Instances          P          R      mAP50        mAP50-95: 100% 12/12 [00:17<00:00,  1.44s/it]
                   all             731       2466         0.549      0.495      0.475      0.227
              Potholes        731        558          0.593      0.579      0.553      0.258
             AligatorC        731        800          0.755      0.696      0.739      0.435
         LongitudanalC    731       791           0.452      0.425      0.353       0.14
           TransverseC       731        317          0.396      0.279      0.254     0.0775

there is a class imbalances for the dataset and instances per class is as follow
pothole: 6000
AligatorC : 11000
longitudanalC: 8000
transverseC:3000
1. How can  I further improve the model performance
2 .The images are resized to 640x640 , but I have images in 3 different resolution. should I train again in the 
   respective native resolution to obtain better results or should I use higher model version?
3. How can this be done
",Always train from a local dataset. Mounted or network drives like Google Drive will be very slow".

**Reply by glenn-jocher** [reply]
Hello,

Here are some suggestions to further improve the performance of your model based on your scenario:

1. **Addressing Class Imbalance**: You might want to try techniques like adjusting the class weights in the loss function or using more advanced sampling methods like SMOTE to balance your dataset during training.

2. **Image Resolution**: Training at native resolution can help if the dataset includes a lot of fine details that are relevant to your classes. However, this would increase computational costs. Given that you're using Google Colab, resources might be limited, so using a stronger version of YOLO, such as YOLOv5m or YOLOv5l, might provide a better improvement while keeping the resolution at 640x640.

3. **Local versus Mounted Dataset**: Training from a local dataset helps speed up the training process as data transfer rates are much higher. If you are bound to use online platforms like Colab, ensure your data is packaged neatly in a way that minimizes I/O time, perhaps using zip files and then unzipping them directly into the environment's local storage.

For additional tips, consider reviewing our detailed guide on achieving the best training results with YOLOv5: [Tips for Best Training Results](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/).

Good luck with your thesis! Let us know if you have more questions. 🚀

**Comment by kishore-work-hard** [comment]
My question is how to quantize a yolo v8 model? Pruning and quantization. Is there any tutorials ?

**Reply by glenn-jocher** [reply]
Currently, YOLOv8 does not natively support model quantization or pruning directly through the provided tools or documentation. For quantization and pruning, you might need to explore third-party libraries or frameworks that support these operations on PyTorch models, such as `torch.quantization` for quantization and `torch.nn.utils.prune` for pruning.

For a basic example of applying quantization in PyTorch, you can start with something like this:

```python
import torch
from ultralytics import YOLO

# Load your model
model = YOLO('yolov8n.pt').to('cpu')

# Specify quantization configuration
model.fuse()  # Fuse Conv2d + BatchNorm2d layers if applicable
model.qconfig = torch.quantization.get_default_qconfig('fbgemm')

# Prepare and convert the model to quantized version
torch.quantization.prepare(model, inplace=True)
torch.quantization.convert(model, inplace=True)
```

This is a basic approach and might need adjustments based on your specific requirements. For more detailed guidance on quantization and pruning, you might need to refer to the PyTorch documentation or other specialized resources. If you have further questions or need more specific advice, feel free to ask!

**Comment by beatsea20** [comment]
Hey, how can I apply data augmentation to increase the size of  dataset before training the model. I was confused about the annotations of those augmented images. Do I have to annotate them again?

**Reply by glenn-jocher** [reply]
@beatsea20 to apply data augmentation to your dataset before training, you can use libraries like Albumentations or TensorFlow's ImageDataGenerator. These tools can generate augmented images along with their corresponding annotations, so you don't need to manually re-annotate them. YOLOv8 also supports data augmentation directly through its configuration file, where you can specify various augmentation techniques. For detailed guidance, please refer to our [data preprocessing guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Reply by beatsea20** [reply]
@glenn-jocher  Thanks, kindly help me with two things.
1. if the model is loaded by 'model = YOLO("yolov8n.yaml")' how do I edit the configuration file?
2. I'm trying to train yolov8n on custom dataset. I tried using both the pretrained model and training from scratch. Both ways the model fails to predict anything on new image.
Since the dataset is small (150 images), I thought using the pretrained model and higher epochs (as suggested by you in an issues on github) but even then there's no predictions by the model.

for training: 
from ultralytics import YOLO
model = YOLO("yolov8n.pt")
results = model.train(data="config.yaml", epochs=100, batch=32, verbose=True)

for prediction:
from ultralytics import YOLO
model = YOLO("\yolo\runs\detect\train4\weights\best.pt")

img_path = "path_to_img"
results = model.predict(source=img_path, save=True)
for result in results:
    result.save_txt('\yolo\runs\detect\train4\output.txt', save_conf=True)
    boxes = result.boxes
    print(boxes) #returns empty lists

**Reply by glenn-jocher** [reply]
To edit the configuration file when loading the model with `YOLO("yolov8n.yaml")`, you can modify the YAML file directly before loading it. Ensure you save the changes before using it in your script.

Regarding your training issue, small datasets like yours (150 images) can be challenging. Using pretrained weights and increasing epochs is a good approach, but also consider data augmentation to artificially increase dataset size and variety. Ensure your annotations are accurate and consistent. If the model still fails to predict, try validating your dataset and training process for any inconsistencies. For more detailed guidance, please refer to our [data preprocessing guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Reply by beatsea20** [reply]
@glenn-jocher I'm using the 'ultralytics' python package, so there's no 'yolov8n.yaml' file in the directory.

**Reply by glenn-jocher** [reply]
If you're using the 'ultralytics' Python package and don't have a 'yolov8n.yaml' file, you can still modify the model configuration directly in your script. For training issues, ensure your dataset is well-annotated and consider using data augmentation to increase its size. If the model fails to predict, verify the annotations and training process. For more detailed guidance, please refer to our [data preprocessing guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Comment by danegarrido** [comment]
If I want to include background images (images without the object of interest) in my custom dataset to fine-tune a YOLO model, should the .txt files containing the labels be empty or not exist at all?

**Reply by glenn-jocher** [reply]
@danegarrido for background images, you simply omit the .txt files—no label file is required when there are no objects to annotate.

**Comment by Cola-shao** [comment]
Why is there no option to save logs in this version now?,Is it saved by default?,But I didn't find it with tensorboard.

**Reply by glenn-jocher** [reply]
@Cola-shao training logs are saved by default in your `runs/train/exp` directory. For TensorBoard logs, ensure you're pointing to the correct path with `tensorboard --logdir runs/train`. If you specified custom `project` or `name` arguments during training, check those directories instead. The Tips for Best Training Results guide shows how to interpret log outputs. Let us know if you need more help locating your files! 🚀

**Reply by Cola-shao** [reply]
Thank you very much! I've fixed this issue.

**Reply by glenn-jocher** [reply]
Training logs are saved by default in `runs/train/exp*`. For TensorBoard, ensure you're running `tensorboard --logdir runs/train` from your project root. If you specified custom `project` or `name` arguments during training, adjust the path accordingly. The Configuration documentation explains log management in detail. Let us know if you need further assistance! 🚀

**Comment by Cola-shao** [comment]
In your train.py code, the learning-rate seem to be set by a dictionary named meta. How can I modify the learning rate when I start, just modify the "lr0" in "meta"?

**Reply by glenn-jocher** [reply]
@Cola-shao yes, to modify the initial learning rate, you can adjust the `lr0` value in your hyperparameters file (e.g., `hyp.scratch.yaml`) or pass `--lr0 [value]` directly in your training command. For full learning rate configuration details, see the Train Settings documentation.

**Comment by Cola-shao** [comment]
@glenn-jocher   I want to modify the model to add CBAM attention method, but when I trained it , an error came out which said  RuntimeError: adaptive_max_pool2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True)'. But in your code, you used `nn.AdaptiveAvgPool2d(1)`  but no error is reported . That's why? Could you please teach me how to add an attention method to improve the model's performance？

**Reply by glenn-jocher** [reply]
@Cola-shao hi! Thanks for your question. To resolve the deterministic error when adding attention modules, you can set `deterministic=False` in your training command. For integrating attention mechanisms into YOLO architectures, we recommend reviewing our Model Modification Guide for best practices. This approach maintains reproducibility while allowing custom layer additions. Let us know if you need further assistance! 🚀

**Comment by Cola-shao** [comment]
What should I set if I want to use mixed precision training. There is no option named --amp in YOLOv5 v6.0. Is it because it is not allowed in this version?

**Reply by glenn-jocher** [reply]
@Cola-shao for YOLOv5 v6.0, mixed precision training is automatically enabled when using compatible hardware (CUDA-capable GPUs). The `--amp` argument was introduced in later versions. To ensure optimal performance in v6.0, verify you're using PyTorch ≥1.7.0 and a compatible GPU. For current implementations, see the Train Settings documentation which explains how AMP works in newer releases.

**Comment by Cola-shao** [comment]
@glenn-jocher What should I do if I want to deploy my own model which is well trained into an ESP32 device. I noticed that there is a Python file named `export.py` which can help me to translate `.pt` into `.onnx` and others, but I don't know how to use it. Could you please help me fix this issue?

**Reply by glenn-jocher** [reply]
@Cola-shao to deploy your trained model to an ESP32 device, you'll need to export it to a format compatible with microcontrollers. You can use the `export.py` script as follows:

```python
# Export using Python
from ultralytics import YOLO
model = YOLO('path/to/your/best.pt')
model.export(format='onnx')  # exports to ONNX format

# Or via command line
yolo export model=path/to/your/best.pt format=onnx
```

For ESP32 specifically, you'll likely need a TFLite or int8 quantized model format for efficient deployment. After exporting to ONNX, you might need additional conversion steps using TensorFlow tools. Consider using the TFLite Micro framework which is designed for microcontrollers like ESP32.

For detailed guidance on optimizing models for edge devices, check our [Export documentation](https://docs.ultralytics.com/modes/export/) which covers various export formats and deployment options.

**Reply by Cola-shao** [reply]
The model I exported into 'tflite' has  about 7MB. but when I transfor it into '.h' file, it almost has 45MB.That's too big for an ESP32 device. Is there any method to fix this problem. How to make the model smaller. I have no idea.

**Reply by glenn-jocher** [reply]
@Cola-shao the size increase from TFLite to .h file is expected - the .h file contains the model as a C array with additional formatting. To reduce your model size for ESP32 deployment, try these approaches:

1. Use post-training quantization to int8 with `model.export(format='tflite', int8=True)`
2. Consider a smaller base model like YOLO11n or YOLO11s
3. Apply pruning to remove less important weights
4. Reduce input resolution (e.g., 320x320 instead of 640x640)
5. Limit the number of classes if possible

For microcontroller deployment guides specific to ESP32, check our [Edge TPU on Raspberry Pi guide](https://docs.ultralytics.com/guides/raspberry-pi/) which includes principles that apply to other edge devices as well.
